### ISSUE TITLE
Implement access-limiting behaviour for Whitehall attachments in Asset Manager

### ISSUE THREAD

**Issue by floehopper** [alphagov/asset-manager#450]
e.g. limited to groups of users in an organisation

**Comment by chrisroos** [comment]
I'm going to start working on this.

**Comment by chrisroos** [comment]
I'm working on the changes necessary to Asset Manager in "WIP: Allow draft asset access to be restricted to specific users #471".

I'm working on the changes necessary to Whitehall in "WIP: Send access_limited data to Asset Manager #3767".

**Comment by chrisroos** [comment]
The changes in https://github.com/alphagov/asset-manager/pull/471 and https://github.com/alphagov/whitehall/pull/3767 have been deployed to production.

@floehopper realised that we also need to handle attachments created by the `BulkUploadsController` in Whitehall. I had hoped that the `BulkUploadsController` might not be used any longer but there are (a small number of) requests for it within the last 14 days so I think we do need to handle these attachments.

**Comment by chrisroos** [comment]
I made the changes to the `BulkUploadsController` in Whitehall in https://github.com/alphagov/whitehall/pull/3822 which has also been merged to master.

I plan to test the functionality in integration to ensure that it's working as expected.

**Comment by floehopper** [comment]
@chrisroos Does this work for `OfficialStatistics` & `NationalStatistics` publication types which are access-limited by default?

* [`PublicationType::OfficialStatistics`][1]
* [`PublicationType::NationalStatistics`][2]

[1]: https://github.com/alphagov/whitehall/blob/32bb13cf3ae3ae9ca71fa14bd4c620c631660419/app/models/publication_type.rb#L100
[2]: https://github.com/alphagov/whitehall/blob/32bb13cf3ae3ae9ca71fa14bd4c620c631660419/app/models/publication_type.rb#L101

**Comment by chrisroos** [comment]
Good question, @floehopper. I'll investigate.

**Comment by chrisroos** [comment]
> @chrisroos Does this work for `OfficialStatistics` & `NationalStatistics` publication types which are access-limited by default?

I've checked and believe that this will just work because setting `Publication#publication_type` to something that's access limited by default will set the `access_limited` attribute of that publication to `true`:

```
> Publication.new.access_limited?
=> nil
> Publication.new(publication_type: PublicationType::OfficialStatistics).access_limited?
=> true
> Publication.new(publication_type: PublicationType::PolicyPaper).access_limited?
=> false
```

**Comment by chrisroos** [comment]
I've just tested this in integration and think I've found a problem. The `access_limited` array of UIDs doesn't appear to be reset to an empty array when an access-limited edition in Whitehall is updated so that it's no longer access-limited. I'm going to investigate further.

**Comment by chrisroos** [comment]
I've opened https://github.com/alphagov/asset-manager/pull/510 to fix the problem that occurs when we try to reset the access-limited state of an asset.

**Comment by chrisroos** [comment]
PR #510 has been merged and deployed to integration and I've now confirmed that this functionality is working as expected.

I created an access-limited consultation in Whitehall integration and added two attachments: one directly and one via the bulk upload functionality. I observed that both attachments were marked as access-limited in Asset Manager. I then removed the access-limiting from the consultation and observed that both attachments were no longer marked as access-limited in Asset Manager.

I'm happy that this is working as expected so I'm closing this issue.
